Yosemite/Yosemite/Model/Mocks/ActionHandlers/MockProductActionHandler.swift

```diff
@@ -0,0 +1,72 @@
+import Foundation
+import Storage
+import Networking
+
+struct MockProductActionHandler: MockActionHandler {
+
+    typealias ActionType = ProductAction
+
+    let objectGraph: MockObjectGraph
+    let storageManager: StorageManagerType
+
+    private let productStore: ProductStore
+
+    init(objectGraph: MockObjectGraph, storageManager: StorageManagerType) {
+        self.objectGraph = objectGraph
+        self.storageManager = storageManager
+
+        productStore = ProductStore(dispatcher: Dispatcher(), storageManager: storageManager, network: NullNetwork())
+    }
+
+    func handle(action: ActionType) {
+        switch action {
+            case .requestMissingProducts(let order, let onCompletion):
+                requestMissingProducts(for: order, onCompletion: onCompletion)
+            case .retrieveProducts(let siteID, let productIDs, _, _, let onCompletion):
+                retrieveProducts(siteId: siteID, productIds: productIDs, onCompletion: onCompletion)
+            case .synchronizeProducts(let siteID, _, _, _, _, _, _, let excludedProductIDs, _, let onCompletion):
+                synchronizeProducts(siteID: siteID, excludedProductIDs: excludedProductIDs, onCompletion: onCompletion)
+            default: unimplementedAction(action: action)
+        }
+    }
+
+    func synchronizeProducts(siteID: Int64, excludedProductIDs: [Int64], onCompletion: @escaping (Result<Bool, Error>) -> Void) {
+        let products = objectGraph.products(forSiteId: siteID, without: excludedProductIDs)
+        upsert(products: products) {
+            /// Indicate that no more products are coming
+            onCompletion(.success(false))
+        }
+    }
+
+    func retrieveProducts(
+        siteId: Int64,
+        productIds: [Int64],
+        onCompletion: @escaping (Result<(products: [Product], hasNextPage: Bool), Error>) -> Void
+    ) {
+        let products = objectGraph.products(forSiteId: siteId, productIds: productIds)
+        upsert(products: products) {
+            onCompletion(.success((products, false)))
+        }
+    }
+
+    func requestMissingProducts(for order: Order, onCompletion: @escaping (Error?) -> Void) {
+        let productIds = order.items.map { $0.productID }.uniqued()
+        let products = objectGraph.products(forSiteId: order.siteID, productIds: productIds)
+
+        upsert(products: products) {
+            onCompletion(nil)
+        }
+    }
+
+    func upsert(products: [Product], onCompletion: @escaping () -> ()) {
+        let storage = storageManager.newDerivedStorage()
+
+        storage.perform {
+            productStore.upsertStoredProducts(readOnlyProducts: products, in: storage)
+        }
+
+        storageManager.saveDerivedType(derivedStorage: storage) {
```

Comment:
@jkmassel Just a quick question, it's probably not super important here either but could we end up a race condition with the perform block here? I'm wondering if _technically_ we could end up saving before the perform block completes

Reply:
Ya this is an excellent point – I can't believe I missed this 🙈 

It's the same code as is used here: 

https://github.com/woocommerce/woocommerce-ios/blob/bdbabd20ca3ee0a581c7c1b1e5c8bafd7a319eeb/Yosemite/Yosemite/Stores/ProductStore.swift#L345-L357


@shiki – WDYT? IMHO this makes sense to fix in both places right?

Reply:
I have been confused by this in the past. This is fine as far as I can tell. The `saveDerivedType` has a `perform` call too. 🙂 So they should happen in sequence. 

https://github.com/woocommerce/woocommerce-ios/blob/bdbabd20ca3ee0a581c7c1b1e5c8bafd7a319eeb/Storage/Storage/CoreData/CoreDataManager.swift#L143-L152

<img src="https://media.giphy.com/media/QMHoU66sBXqqLqYvGO/giphy.gif" width="300">